### PR TITLE
Improve SAT resolving developer debug information

### PR DESCRIPTION
**Pull request by Toflar**
Currently, when you debug using `-vvv`, with rather complex dependencies, your CLI output is stuck at `Resolving dependencies through SAT` for 20+ seconds and you don't know if Composer is still doing anything or it crashed, so I tried to improve this.

Before it just prompts after 22 seconds:

bildschirmfoto 2018-05-18 um 11 03 05

After:

ezgif com-video-to-gif

### CODE REVIEW

File: src/Composer/DependencyResolver/Solver.php [hunk 1]
```diff
@@ -762,6 +763,7 @@ private function runSat($disableRules = true)
 
             for ($i = 0, $n = 0; $n < $rulesCount; $i++, $n++) {
                 if ($i == $rulesCount) {
+                    $this->io->writeError('.', false, IOInterface::DEBUG);
```
Comment (staabm) [inline src/Composer/DependencyResolver/Solver.php]: I would love a single space before all those dots get printed

**Comment by curry684** [comment]
When I added the "Resolving dependencies through SAT" log line I did that specifically because it isolates a single line of code that tends to run for a long time. I'm not sure the dots really help, I mean it's SAT so it cannot hang indefinitely by definition. I tried to find some more reliable way of displaying progress but didn't find one at the time (again, because SAT pretty much by definition does not know when it will find a solution).

**Comment by Toflar** [comment]
I've placed it exactly at the place where a re-run happens. This happens like 20 times for my ruleset. So if you have a smaller one, likely there will be less re-runs and if you have more, it'll likely re-run more often. So it's a pretty accurate progress indicator although it will print dots faster towards the end because the number of comparisons needed gets smaller.
Yes, you cannot indicate anything like a percentage because you don't know how long it'll take in advance. And it doesn't have to be reliable at all. It's just an indicator that it's still running (and remember, it's only on debug level, so it won't show unless you pass `-vvv`).

**Comment by alcohol** [comment]
I feel like the dots do not represent anything useful and merely clutter the output. If you're at the debug level already, at least output something debug worthy.

**Comment by Toflar** [comment]
Like so?

```
Resolving dependencies through SAT
Something's changed, looking at all rules again (pass #1)
Something's changed, looking at all rules again (pass #2)
Something's changed, looking at all rules again (pass #3)
[...]
Something's changed, looking at all rules again (pass #20)
Finished resolving dependencies through SAT
```

**Comment by alcohol** [comment]
Sure, though you could overwrite the previous output, rather than outputting a new line again and again.

**Comment by Toflar** [comment]
Here we go, now it looks like so:

ezgif com-optimize

File: src/Composer/DependencyResolver/Solver.php [hunk 2]
```diff
@@ -759,10 +760,19 @@ private function runSat($disableRules = true)
             }
 
             $rulesCount = count($this->rules);
+            $pass = 1;
 
+            $this->io->writeError('Looking at all rules.', true, IOInterface::DEBUG);
             for ($i = 0, $n = 0; $n < $rulesCount; $i++, $n++) {
                 if ($i == $rulesCount) {
+                    if (1 === $pass) {
+                        $this->io->writeError("Something's changed, looking at all rules again (pass #$pass)", false, IOInterface::DEBUG);
```
Comment (staabm) [inline src/Composer/DependencyResolver/Solver.php]: IMO it shoud just be „Looking at all rules (pass #$pass)“.

No need for this more verbose (and non obvious) „something changed“ thing.

Comment (Toflar) [reply src/Composer/DependencyResolver/Solver.php]: I just took the comment there as an idea so. But yeah, whatever. Not gonna change this now, it can be changed to whatever the maintainers prefer.

**Comment by curry684** [comment]
Just a thought - why the hell are we only going to do this on `-vvv`? It would appear to me the ultimate extra line of output *always* to have a `Resolving dependencies...` message which occasionally adds a dot, and will likely be just about as long the next time you run it, making it a semi-reliable indicator on the project level for everyone.

**Comment by alcohol** [comment]
I'm sorry, but can you rephrase that? I'm not really sure what you are saying or asking right now.

**Comment by curry684** [comment]
We've had tons of issues past years, complaining about Composer hanging indefinitely. This PR adds a progress indicator showing it's not dead or hanging. Why are we hiding it in `-vvv` instead of always in plain sight?

**Comment by alcohol** [comment]
Fair point, I'll also leave that up to @Seldaek to decide though :d

**Comment by Seldaek** [comment]
@naderman I think this loop would become redundant with Composer 2.0 refactorings as all providers would load first right? Or is this mistaken? Because if that's the case then I'd rather not merge as it'll become useless.

**Comment by Toflar** [comment]
I don't think it's related. The output is during the recomputation in case any rule was marked as false or true. So it's SAT internal and not related to any previous stage.

**Comment by Seldaek** [comment]
Alrighty then, thanks :)

**Comment by Toflar** [comment]
Cheers, see you later :)
